### PR TITLE
[libc++][CI] Tests the no RTTI configuration.

### PR DESCRIPTION
**Pull request by mordante**
There are a few drive-by fixes:
- Since the combination RTTI disabled and exceptions enabled do not work, this combination is prohibited.
- A small NFC in any fixing clang-tidy.

The code in the Buildkite configuration is prepared for using the std module. There are more fixes needed for that configuration which will be done in a separate commit.

### CODE REVIEW

File: libcxx/CMakeLists.txt [hunk 1]
```diff
@@ -279,7 +279,9 @@ endif()
 
 # Feature options -------------------------------------------------------------
 option(LIBCXX_ENABLE_EXCEPTIONS "Use exceptions." ON)
-option(LIBCXX_ENABLE_RTTI "Use run time type information." ON)
+option(LIBCXX_ENABLE_RTTI
+  "Use run time type information.
+   This option may only be set to OFF when LIBCXX_ENABLE_EXCEPTIONS=OFF." ON)
```
Comment (philnik777) [inline libcxx/CMakeLists.txt:284]: Is this only a problem within libc++ or is there some fundamental reason we can't do this (i.e. is RTTI always required for EH)?

**Review by philnik777 — APPROVED** [review]
LGTM assuming the CI passes.

**Review by EricWF — CHANGES_REQUESTED** [review]
I still have remaining questions.

**Comment by EricWF** [comment]
Sorry, still getting used to github PR's. I thought I had submitted these comments last week.

File: libcxx/CMakeLists.txt [hunk 2]
```diff
@@ -370,6 +372,11 @@ if (LIBCXX_HAS_PTHREAD_API)
   endif()
 endif()
 
+if (NOT LIBCXX_ENABLE_RTTI AND LIBCXX_ENABLE_EXCEPTIONS)
+  message(FATAL_ERROR "The option LIBCXX_ENABLE_RTTI can not be turned off"
```
[the remark this answers is not in the document]
Comment (ldionne) [reply libcxx/CMakeLists.txt:376]: I summarized issues we have with `-fno-rtti -fexceptions` here: https://github.com/llvm/llvm-project/issues/66117.

I would disagree that this is a "well supported" configuration. We've had multiple user reports of things not behaving properly in that configuration, and I would not recommend anybody to use it in its current state.

File: libcxx/utils/ci/buildkite-pipeline.yml [hunk 3]
```diff
@@ -741,6 +741,27 @@ steps:
...
+      artifact_paths:
+        - "**/test-results.xml"
+        - "**/*.abilist"
+      env:
+          # Note: Modules require and absolute path for clang-scan-deps
+          # https://github.com/llvm/llvm-project/issues/61006
+          CC: "/usr/lib/llvm-${LLVM_HEAD_VERSION}/bin/clang"
+          CXX: "/usr/lib/llvm-${LLVM_HEAD_VERSION}/bin/clang++"
+          CMAKE: "/opt/bin/cmake"
+          ENABLE_CLANG_TIDY: "On"
```
[the remark this answers is not in the document]
Comment (mordante) [reply libcxx/utils/ci/buildkite-pipeline.yml:755]: Not per se. A codepath `#ifdef`'ed for this configuration my have issues. So this my find additional issues.

File: libcxx/CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (mordante) [reply libcxx/CMakeLists.txt:376]: Thanks for the great summary @ldionne .

File: libcxx/CMakeLists.txt (hunk 1 above)
Comment (mordante) [reply libcxx/CMakeLists.txt:284]: See #66117 which @ldionne just created.

File: libcxx/include/any [hunk 4]
```diff
@@ -196,9 +196,7 @@ namespace __any_imp
       if (__id && *__id == typeid(_Tp))
           return true;
 #endif
-      if (!__id && __fallback_id == __any_imp::__get_fallback_typeid<_Tp>())
-          return true;
-      return false;
+      return !__id && __fallback_id == __any_imp::__get_fallback_typeid<_Tp>();
```
Comment (ldionne) [inline libcxx/include/any:199]: Is there a reason why this is part of this patch?

**Review by ldionne — APPROVED** [review]
LGTM, this adds important missing coverage to our CI.

I would not make the changes in `libcxx/include/any` in this PR though, it should be a separate NFC IIUC.

File: libcxx/include/any (hunk 4 above)
Comment (mordante) [reply libcxx/include/any:199]: Yes the removal of the `#if !defined(_LIBCPP_HAS_NO_RTTI)` block gives a clang-tidy warning.

**Comment by mordante** [comment]
@EricWF are you happy with the changes?

**Review by ldionne — APPROVED** [review]
LGTM with small comments.

**Comment by ldionne** [comment]
> I would much rather we do what our users tend to do here, and that's build with rtti enabled, and then disable rtti only during the test suite run.

In reality, we have instances of both things happening. Most embedded configurations will actually build the library without RTTI. I wonder if we really want to test both combinations -- we generally don't.

I do agree that most users on e.g. desktops will disable RTTI yet their dylib will have been built with RTTI enabled. I think there's a question of how we want to approach this configuration explosion. FWIW I could even be talked into dropping support for building the library without RTTI -- I wonder how much impact this actually has on the size of the library but I suspect not that much.

**Comment by mordante** [comment]
The reason for adding the configuration is that users can do this and it was not tested.

Maybe we should brainstorm how we can avoid running all configurations for all patches, but still allow to run them manually. For example, format related patches really should test no-wide-characters, mdspan patches are very unlikely to find an issue in that configuration.

**Review by ldionne — APPROVED** [review]
LGTM -- this adds coverage for something we claim to support right now but didn't test. I'd be in favour of investigating whether we actually want to support building with RTTI disabled at all (perhaps not), but that would be a separate effort.

**Pull request merged**